projects/ccf-rui/src/app/shared/components/visibility-menu/visibility-menu.component.ts

```diff
@@ -0,0 +1,116 @@
+import { Component, Input, Output, EventEmitter, HostBinding } from '@angular/core';
+
+/**
+ * Interface for visibility item data
+ */
+export interface VisibilityItem {
+
+  /**
+   * Id of the item
+   */
+  id: number;
+
+  /**
+   * Name of the item
+   */
+  name: string;
+
+  /**
+   * Whether the item is currently highlighted
+   */
+  visible: boolean;
+
+  /**
+   * Src of the icon (visible or non-visible)
+   */
+  iconSrc: string;
+
+  /**
+   * Opacity value
+   */
+  opacity?: number;
+
+  /**
+   * Tooltip text to be displayed in the stage
+   */
+  tooltip?: string;
+}
+
+/**
+ * Menu for displaying visibility options
+ */
+@Component({
+  selector: 'ccf-visibility-menu',
+  templateUrl: './visibility-menu.component.html',
+  styleUrls: ['./visibility-menu.component.scss']
+})
+export class VisibilityMenuComponent {
+
+  /**
+   * HTML class name
+   */
+  @HostBinding('class') readonly clsName = 'ccf-visibility-menu';
+
+  /**
+   * Items to be displayed in the visibility menu
+   */
+  @Input() items: VisibilityItem[];
+
+  /**
+   * Items that are currently set to visible
+   */
+  @Input() visibleItems: VisibilityItem[];
+
+  /**
+   * Whether opacity value should be displayed
+   */
+  @Input() opacityOn = false;
+
+  /**
+   * Emits the currently visible items
+   */
+  @Output() visibleItemsChange = new EventEmitter<VisibilityItem[]>();
+
+  /**
+   * Emits the currently selected item
+   */
+  @Output() selectionChange = new EventEmitter<VisibilityItem | undefined>();
+
+  /**
+   * Emits the currently hovered item
+   */
+  @Output() hover = new EventEmitter<VisibilityItem | undefined>();
+
+  /**
+   * The currently selected item
+   */
+  selectedItem: VisibilityItem | undefined;
+
+  /**
+   * Toggles highlight state, sets the icon type, and emits an array containing the currently visible items
+   * @param item Menu item
+   */
+  toggleVisibility(item: VisibilityItem): void {
+    item.visible = !item.visible;
+    this.visibleItems = this.items.filter(x => x.visible);
+    item.iconSrc = item.visible ? 'app:visibility_on' : 'app:visibility_off';
+    this.visibleItemsChange.emit(this.visibleItems);
+  }
+
+  /**
+   * Toggles selected status of an item on click and sets visibility status of the item
+   * @param item Menu item
+   */
+  toggleSelected(item: VisibilityItem): void {
+    this.selectedItem = item === this.selectedItem ? undefined : item;
+    this.selectionChange.emit(item);
+  }
+
+  /**
+   * Emits an item in response to hover action
+   * @param item Menu item
+   */
+  mouseOver(item: VisibilityItem): void {
```

Comment:
Don't forget mouse out to `this.hover.emit(undefined);` so they know the hover is over.